### PR TITLE
2.0 base changes

### PR DESCRIPTION
**Pull request by michaelengland**
So. The idea behind this was to get cracking on some of the bigger changes that were discussed with @glung . I ended up losing about a month due to some other stuff but here it is in it's current format. Make with it what you will. Happy to help if I can.

Some things about the PR:
 - Separates Java & Android-only components (`lightcycle-lib` & `lightcycle-android-lib`. This includes adding an example that only uses base Java. I figured this is good for modularization and also allows a slightly improved array of tools to be used.
 - Fixed bug with base classes [incorrectly swallowing item selection](https://github.com/soundcloud/lightcycle/issues/53).
 - Fixes bug with base classes only accepting certain subtypes in light cycles.
 - Removes base classes, instead favoring the newly introduced `@LightCycleBaseClass`. This annotation can be put on any abstract fragment/support fragment/activity that is said to implement a `LightCycleDispatcher`. From there, the `lightcycle-android-processor` will auto-generate a `LightCycle_<Original>` base class that dispatches all the methods correctly.
 - Adds some missing test cases that were then fixed.
 - Tidied test code up a bit (i.e. symbolic link for the processor tests `lightcycle-android-lib` source folder in order to lead to fewer mismatches between code & tests.

Still todo:
 - [ ] Get new `lightcycle-android-processor` working correctly with `lightcycle-processor`. Right now, both work independently, but annoyingly the `lightcycle-processor` believes that the auto-generated class is erroring. Any help on this would be especially appreciated, as I have really no idea what I'm doing with all the processors.
 - [ ] Re-introduce lifts (removed for now otherwise couldn't separate Java & Android-only in its previous state & was also causing problems with user created base classes).
 - [ ] Not up-to-date with master (e.g. need to improve tests now given recent huge improvements in gradle & testing setup).
 - [ ] Add all system lifecycle & user-triggered lifecycle methods, in order to cover as many potential library user cases as possible.

Also, ofc, any other feedback/suggestions/hatred of what I've done is appreciated. :)

### CODE REVIEW

**Comment by NikoYuwono** [comment]
Nice implementation! 

If I may ask, is there any other special reason to separate lightcycle-lib into Java and Android-only component beside modularization?  
I saw your example (race one) and figured out it's a dispatcher for classes that implemented some interface and while I think it's also a good idea, isn't it kinda different from the main objective of Lightcycle that helps break logic out of `Activity` and `Fragment `classes into small, self-contained components? 

I think I can help with the processor errors if the plan is still to separate into Java and Android-only component and also adding missing lifecycle methods! 😃

File: examples/basic/src/main/java/com/soundcloud/lightcycle/sample/basic/BaseActivity.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import android.support.v7.app.AppCompatActivity;
+
+import com.soundcloud.lightcycle.ActivityLightCycle;
+import com.soundcloud.lightcycle.LightCycleBaseClass;
+import com.soundcloud.lightcycle.LightCycleDispatcher;
+
+@LightCycleBaseClass
+public abstract class BaseActivity extends AppCompatActivity
+        implements LightCycleDispatcher<ActivityLightCycle<AppCompatActivity>> {
```
Comment (glung) [inline examples/basic/src/main/java/com/soundcloud/lightcycle/sample/basic/BaseActivity.java]: Do you think we can have a way to specify the Host's the of the dispatcher? 

()https://github.com/soundcloud/lightcycle/issues/48)
> Host : the instance of the Activity or Fragment's lifecycle as seen by the LightCycle.
> The host can be an Activity, a Fragment or anything else (can be an interface like MyView).

Examples :

```
public abstract class BaseActivity extends AppCompatActivity
         implements LightCycleDispatcher<ActivityLightCycle<BaseActivity>> {
```

```
public abstract class BaseActivity extends AppCompatActivity
          implements MyView, LightCycleDispatcher<ActivityLightCycle<MyView>> {
```

File: lightcycle-android-processor/src/main/java/com/soundcloud/lightcycle/AbstractFragmentClassWriter.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                .addMethod(onResumeMethod())
+                .addMethod(onOptionsItemSelectedMethod())
+                .addMethod(onPauseMethod())
+                .addMethod(onStopMethod())
+                .addMethod(onSaveInstanceStateMethod())
+                .addMethod(onDestroyViewMethod())
+                .addMethod(onDestroyMethod())
+                .addMethod(onDetachMethod())
+                .addMethod(objectMethod())
+                .addMethod(bindIfNecessaryMethod())
```
Comment (glung) [inline lightcycle-android-processor/src/main/java/com/soundcloud/lightcycle/AbstractFragmentClassWriter.java]: Yo dawg !

File: examples/basic/src/main/java/com/soundcloud/lightcycle/sample/basic/BaseActivity.java (hunk 1 above)
Comment (glung) [inline examples/basic/src/main/java/com/soundcloud/lightcycle/sample/basic/BaseActivity.java]: Actually, looking more at the tests, the answer seems to be yes : 

```
@LightCycleBaseClass
public abstract class BaseActivityAsObject<T extends String> extends Activity
        implements LightCycleDispatcher<ActivityLightCycle<BaseActivityAsObject<T>>> {
}
```

do you confirm ?

Comment (michaelengland) [inline examples/basic/src/main/java/com/soundcloud/lightcycle/sample/basic/BaseActivity.java]: Yes. I made sure that it works with things like generics too (as you can see with that test)

**Pull request closed without merging**